grails-app/controllers/org/pih/warehouse/notification/NotificationController.groovy

```diff
@@ -154,5 +159,200 @@ class NotificationController {
         }
     }
 
+    def subscribeOrder = {
+        String ORDER_TOPIC_ARN = grailsApplication.config.awssdk.sns.order.arn?.toString()
+
+        log.info "Subscribiing topic:${ORDER_TOPIC_ARN}"
+        try {
+            String subscribeUrl = grailsApplication.config.grails.serverURL + "/api/notifications/orders"
+            println "subscribeUrl::${subscribeUrl}"
+            SubscribeRequest subscribeRequest = new SubscribeRequest(ORDER_TOPIC_ARN, "https", subscribeUrl)
+            subscribeRequest.returnSubscriptionArn = true
+            SubscribeResult result = amazonSnsClient.subscribe(subscribeRequest);
+            log.info("Subscription ARN is " + result.subscriptionArn);
+        } catch (Exception e) {
+            log.error(e.printStackTrace());
+        }
+        render "subscribed"
+    }
+
+    def createOrder = {
+        JSONObject json = request.getJSON() as JSONObject
+        log.info "Create order JSON:${json}"
+        if (json.has("Type") && json.getString("Type") == "SubscriptionConfirmation") {
+            String token = json.getString("Token")
+            try {
+                String ORDER_TOPIC_ARN = grailsApplication.config.awssdk.sns.order.arn
+                ConfirmSubscriptionRequest request = new ConfirmSubscriptionRequest(ORDER_TOPIC_ARN, token)
+                ConfirmSubscriptionResult result = amazonSnsClient.confirmSubscription(request);
+                log.info "result::${result?.toString()}"
+            } catch (Exception e) {
+                log.error "Error in saving order: " + e.message, e
+            }
+            render "subscribed"
+            return
+        }
+        String message = json.getString("Message")
+        JSONObject oJson = new JSONObject(message)
+//        JSONArray orders = productsJson.getJSONArray("orders")
+//        for(int i = 0;i<orders.length();i++){
+//            JSONObject oJson = orders.getJSONObject(i)
+        log.info "ORDER JSON ${oJson}"
+        Requisition requisition = new Requisition()
+        if (oJson.has("name")) {
+            requisition.name = oJson.getString("name")
+        } else {
+            requisition.name = "DEFAULT"
+        }
+        try {
+            if (oJson.has("origin") && oJson.getString("origin")) {
+                JSONObject jsonObject = oJson.getJSONObject("origin")
+                log.info "Origin location Json:${jsonObject}"
+                Location location = Location.findByLocationNumber(jsonObject.getString("id"))
+                if (!location && grailsApplication.config?.openboxes?.integration?.destinationMapping) {
+                    log.info "Location Mapping::${grailsApplication.config?.openboxes?.integration?.destinationMapping}"
+                    JSONObject destinationMapping = toJson(grailsApplication.config?.openboxes?.integration?.destinationMapping)
```

Comment:
We should at least break this out into a separate method so we don't repeat the code in both places.
```
getLocationByLocationMapping(jsonObject.getString("id")
```